File: sdks/python/src/echo_agent_sdk/memory_policy_values.py
from __future__ import annotations

from enum import Enum


class MemoryType(str, Enum):
    USER_PREFERENCE = "user_preference"
    PROJECT_FACT = "project_fact"
    ARCHITECTURE_DECISION = "architecture_decision"
    DEBUGGING_LESSON = "debugging_lesson"
    ERROR_RESOLUTION = "error_resolution"
    COMMAND_PATTERN = "command_pattern"
    TOOL_USAGE = "tool_usage"
    WORKFLOW_PATTERN = "workflow_pattern"
    SKILL_CANDIDATE = "skill_candidate"
    DEPRECATED_NOTE = "deprecated_note"


class MemorySource(str, Enum):
    USER_CORRECTION = "user_correction"
    ERROR_RESOLUTION = "error_resolution"
    REPEATED_WORKFLOW = "repeated_workflow"
    EXPLICIT_SAVE = "explicit_save"
    AUTO_EXTRACTED = "auto_extracted"
    L3_PROMOTION = "l3_promotion"


_TYPE_STABILITY = {
    MemoryType.USER_PREFERENCE: 0.85,
    MemoryType.PROJECT_FACT: 0.60,
    MemoryType.ARCHITECTURE_DECISION: 0.80,
    MemoryType.DEBUGGING_LESSON: 0.55,
    MemoryType.ERROR_RESOLUTION: 0.50,
    MemoryType.COMMAND_PATTERN: 0.40,
    MemoryType.TOOL_USAGE: 0.50,
    MemoryType.WORKFLOW_PATTERN: 0.60,
    MemoryType.SKILL_CANDIDATE: 0.55,
    MemoryType.DEPRECATED_NOTE: 0.10,
}
_SOURCE_CONFIDENCE = {
    MemorySource.USER_CORRECTION: 0.90,
    MemorySource.ERROR_RESOLUTION: 0.85,
    MemorySource.REPEATED_WORKFLOW: 0.75,
    MemorySource.EXPLICIT_SAVE: 1.0,
    MemorySource.AUTO_EXTRACTED: 0.60,
    MemorySource.L3_PROMOTION: 0.50,
}
_SOURCE_RECALL_WEIGHT = {
    MemorySource.USER_CORRECTION: 0.90,
    MemorySource.ERROR_RESOLUTION: 0.70,
    MemorySource.REPEATED_WORKFLOW: 0.60,
    MemorySource.EXPLICIT_SAVE: 0.80,
    MemorySource.AUTO_EXTRACTED: 0.40,
    MemorySource.L3_PROMOTION: 0.40,
}


def memory_type_default_stability(memory_type: MemoryType) -> float:
    return _TYPE_STABILITY[_validate_type(memory_type)]


def memory_type_is_skill_eligible(memory_type: MemoryType) -> bool:
    return _validate_type(memory_type) in {
        MemoryType.WORKFLOW_PATTERN,
        MemoryType.SKILL_CANDIDATE,
        MemoryType.DEBUGGING_LESSON,
    }


def memory_type_is_rule_eligible(memory_type: MemoryType) -> bool:
    return _validate_type(memory_type) in {
        MemoryType.USER_PREFERENCE,
        MemoryType.ARCHITECTURE_DECISION,
        MemoryType.PROJECT_FACT,
    }


def memory_source_default_confidence(source: MemorySource) -> float:
    return _SOURCE_CONFIDENCE[_validate_source(source)]


def memory_source_default_recall_weight(source: MemorySource) -> float:
    return _SOURCE_RECALL_WEIGHT[_validate_source(source)]


def _validate_type(value: MemoryType) -> MemoryType:
    if not isinstance(value, MemoryType):
        raise TypeError(f"unknown memory type: {value!r}")
    return value


def _validate_source(value: MemorySource) -> MemorySource:
    if not isinstance(value, MemorySource):
        raise TypeError(f"unknown memory source: {value!r}")
    return value

File: sdks/typescript/src/memory_policy_values.ts
/** Typed memory classifications and source policies projected from Rust. */
export const MemoryType = Object.freeze({
  UserPreference: "user_preference",
  ProjectFact: "project_fact",
  ArchitectureDecision: "architecture_decision",
  DebuggingLesson: "debugging_lesson",
  ErrorResolution: "error_resolution",
  CommandPattern: "command_pattern",
  ToolUsage: "tool_usage",
  WorkflowPattern: "workflow_pattern",
  SkillCandidate: "skill_candidate",
  DeprecatedNote: "deprecated_note",
} as const);

export type MemoryType = (typeof MemoryType)[keyof typeof MemoryType];

export const MemorySource = Object.freeze({
  UserCorrection: "user_correction",
  ErrorResolution: "error_resolution",
  RepeatedWorkflow: "repeated_workflow",
  ExplicitSave: "explicit_save",
  AutoExtracted: "auto_extracted",
  L3Promotion: "l3_promotion",
} as const);

export type MemorySource = (typeof MemorySource)[keyof typeof MemorySource];

const TYPE_STABILITY: Readonly<Record<MemoryType, number>> = Object.freeze({
  user_preference: 0.85,
  project_fact: 0.6,
  architecture_decision: 0.8,
  debugging_lesson: 0.55,
  error_resolution: 0.5,
  command_pattern: 0.4,
  tool_usage: 0.5,
  workflow_pattern: 0.6,
  skill_candidate: 0.55,
  deprecated_note: 0.1,
});
const SOURCE_CONFIDENCE: Readonly<Record<MemorySource, number>> = Object.freeze({
  user_correction: 0.9,
  error_resolution: 0.85,
  repeated_workflow: 0.75,
  explicit_save: 1,
  auto_extracted: 0.6,
  l3_promotion: 0.5,
});
const SOURCE_RECALL_WEIGHT: Readonly<Record<MemorySource, number>> = Object.freeze({
  user_correction: 0.9,
  error_resolution: 0.7,
  repeated_workflow: 0.6,
  explicit_save: 0.8,
  auto_extracted: 0.4,
  l3_promotion: 0.4,
});
const MEMORY_TYPES: ReadonlySet<string> = new Set(Object.values(MemoryType));
const MEMORY_SOURCES: ReadonlySet<string> = new Set(Object.values(MemorySource));

export function memoryTypeDefaultStability(memoryType: MemoryType): number {
  validate(memoryType, MEMORY_TYPES, "memory type");
  return TYPE_STABILITY[memoryType];
}

export function memoryTypeIsSkillEligible(memoryType: MemoryType): boolean {
  validate(memoryType, MEMORY_TYPES, "memory type");
  return memoryType === MemoryType.WorkflowPattern
    || memoryType === MemoryType.SkillCandidate
    || memoryType === MemoryType.DebuggingLesson;
}

export function memoryTypeIsRuleEligible(memoryType: MemoryType): boolean {
  validate(memoryType, MEMORY_TYPES, "memory type");
  return memoryType === MemoryType.UserPreference
    || memoryType === MemoryType.ArchitectureDecision
    || memoryType === MemoryType.ProjectFact;
}

export function memorySourceDefaultConfidence(source: MemorySource): number {
  validate(source, MEMORY_SOURCES, "memory source");
  return SOURCE_CONFIDENCE[source];
}

export function memorySourceDefaultRecallWeight(source: MemorySource): number {
  validate(source, MEMORY_SOURCES, "memory source");
  return SOURCE_RECALL_WEIGHT[source];
}

function validate(value: unknown, values: ReadonlySet<string>, name: string): asserts value is string {
  if (typeof value !== "string" || !values.has(value)) throw new TypeError(`unknown ${name}: ${String(value)}`);
}
